docs/dev_guide/contents/extending_fido.rst

```diff
@@ -0,0 +1,501 @@
+
+* ``(a.Instrument("AIA") & a.Time("2020/02/02", "2020/02/03")) | (a.Instrument("HMI") & a.Time("2020/02/02", "2020/02/03"))``
+* ``a.Time("2020/02/02", "2020/02/03")``
+* ``a.Instrument("AIA") & a.Time("2020/02/02", "2020/02/03")``
+* ``(a.Time(..) & a.Instrument("AIA") & a.Wavelength(30*u.nm, 31*u.nm)) | (a.Time(..) & a.Instrument("AIA") & a.Wavelength(30*u.nm, 31*u.nm))``
+
+but you **would not** be passed queries which look like the following examples, even if that's how the user specified them:
+
+* ``a.Time("2020/02/02", "2020/02/03") & (a.Instrument("AIA") | a.Instrument("HMI"))``
+* ``a.Time(..) & (a.Instrument("AIA") | a.Instrument("AIA")) & a.Wavelength(30*u.nm, 31*u.nm))``
+
+The Attr Walker
+###############
+
+Given the potential complexity of these combined attrs, converting them into other forms, such as query parameters or JSON etc involves walking the tree and converting each attr to the expected format in a given way.
+This parsing and conversion of the query tree is deliberately not done using methods or attributes of the attrs themselves.
+The attrs should be independent of any client in their implementation, so they can be shared between the different ``Fido`` clients.
+
+A class is provided to facilitate this conversion, `~sunpy.net.attr.AttrWalker`.
+The `~sunpy.net.attr.AttrWalker` class consists of three main components:
+
+* **Creators**: The `~sunpy.net.attr.AttrWalker.create` method is one of two generic functions for which a different function is called for each Attr type.
+  The intended use for creators is to return a new object dependant on different attrs.
+  It is commonly used to dispatch on `~sunpy.net.attr.AttrAnd` and `~sunpy.net.attr.AttrOr`.
+
+* **Appliers**: The `~sunpy.net.attr.AttrWalker.apply` method is the same as `~sunpy.net.attr.AttrWalker.create` in that it is a generic function.
+  The only difference between it and `~sunpy.net.attr.AttrWalker.create` is its intended use.
+  Appliers are generally used to modify an object returned by a creator with the values or information contained in other Attrs.
+
+* **Converters**: Adding a converter to the walker adds the function to both the creator and the applier.
+  For the VSO client this is used to convert each supported attr into a `~sunpy.net.attr.ValueAttr` which is then later processed by the appliers and creators.
+  This pattern can be useful if you would otherwise have to repeat a lot of logic in each of the applier functions for each type of Attr you support.
+
+An Example of ``AttrWalker``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this example we will write a parser for some simple queries which uses `~sunpy.net.attr.AttrWalker` to convert the query to a `dict` of URL query parameters for a HTTP GET request.
+Let's imagine we have a web service which you can do a HTTP GET request to ``https://sfsi.sunpy.org/search`` for some imaginary data from an instrument called SFSI (SunPy Fake Solar Instrument).
```

Comment:
haha SFSI! 🌞

Reply:
This isn't a real instrument acronym right? :rofl: